rpc.go

```diff
@@ -293,11 +295,104 @@ func (l *Libvirt) deregister(id uint32) {
 // returns response returned by server.
 // if response is not OK, decodes error from it and returns it.
 func (l *Libvirt) request(proc uint32, program uint32, payload []byte) (response, error) {
+	return l.requestStream(proc, program, payload, nil, nil)
+}
+
+func (l *Libvirt) requestStream(proc uint32, program uint32, payload []byte, outStream io.Reader, inStream io.Writer) (response, error) {
 	serial := l.serial()
 	c := make(chan response)
 
 	l.register(serial, c)
+	defer l.deregister(serial)
+
+	err := l.sendPacket(serial, proc, program, payload, Call, StatusOK)
+	if err != nil {
+		return response{}, err
+	}
+
+	resp, err := l.getResponse(c)
+	if err != nil {
+		return resp, err
+	}
+
+	var abortOutStream chan bool
+	var outStreamErr chan error
+	if outStream != nil {
+		abortOutStream = make(chan bool)
+		outStreamErr = make(chan error)
+		go func() {
+			outStreamErr <- l.sendStream(serial, proc, program, outStream, abortOutStream)
+		}()
+	}
+
+	// Even without incoming stream server sends confirmation once all data is received
+	if outStream != nil || inStream != nil {
+		for {
+			resp, err = l.getResponse(c)
+			if err != nil {
+				if outStream != nil {
+					abortOutStream <- true
+				}
+				return resp, err
+			}
+			// Continue is valid here only for stream packets
+			if resp.Status == StatusContinue {
+				if inStream != nil {
+					_, err = inStream.Write(resp.Payload)
+					if err != nil {
+						if outStream != nil {
+							abortOutStream <- true
+						}
+						return response{}, err
+					}
+				}
+			} else {
+				// StatusError is handled in getResponse, so this is StatusOK, end of stream
+				break
+			}
+		}
+	}
+
+	if outStream != nil {
+		err = <-outStreamErr
+		if err != nil {
+			return response{}, nil
+		}
+	}
+
+	return resp, nil
+}
+
+func (l *Libvirt) sendStream(serial uint32, proc uint32, program uint32, stream io.Reader, abort chan bool) error {
+	buf := make([]byte, 1<<22-24)
```

Comment:
what's going on here?

Reply:
Just making sure packet with header won't exceed 4 Mb. I've seen such limit in constants in libvirt and I'm not sure how it works, but I think it's OK to use 4 Mb as a chunk size.

Reply:
That's definitely worthy of a comment. Also I don't like using bit shifting to calculate consts - it's not good for readability. Normally I'd do `4 * units.MiB`, but go-libvirt doesn't have the units package at the moment.

Finally, if the `-24` is really needed, can it be expressed as `len(something)`? If it's not really needed, it should be dropped.

Reply:
Yes, I should've made this line more understandable from the beginning. Fixed now.